Create cron to aggregate old data

mimurawil opened the pull request:
## Description

This creates 3 new cron jobs (and 3 new commands) to aggregate daily data into monthly data:
- Cron: `MonthlyLinkAggregatesCron` / command: `fill_monthly_link_aggregates`
- Cron: `MonthlyUserAggregatesCron` / command: `fill_monthly_user_aggregates`
- Cron: `MonthlyPageProjectAggregatesCron` / command: `fill_monthly_pageproject_aggregates`

This also fixes impacted UI filter and adds a new healthcheck route to provide the status of the monthly jobs.

### Changes

- Create one new index in each `aggregates_linkaggregate`, `aggregates_useraggregate`, and `aggregates_pageprojectaggregate` tables to speed up the query in the new cron jobs
- Setup 3 new cron jobs - running monthly on the 3rd day of the month
- Added new route for `healthcheck` for monthly cron jobs status
- Start and end date filters in `programs` and `organisations` are now year/month picker
- Added tests

## Rationale

We don't need granular data for all the collections by day and some tables (especially pageproject aggregates are getting too big).

## Phabricator Ticket

https://phabricator.wikimedia.org/T370977

## How Has This Been Tested?

Automated tests have been added for
- All 3 new commands
- Custom form field (year/month picker)
- Form clean method

## Screenshots of your changes (if appropriate):
n/a

## Types of changes
What types of changes does your code introduce? Add an `x` in all the boxes that apply:
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

katydidnot reviewed (commented):
This isn't a full review yet as I'm waiting on the data dump to download myself, but wanted to get you what feedback I had.

mimurawil commented on extlinks/aggregates/cron.py:
```diff
@@ -15,6 +15,27 @@ class LinkAggregatesCron(CronJobBase):
     def do(self):
         call_command("fill_link_aggregates")
 
+class MonthlyLinkAggregatesCron(CronJobBase):
```
The daily jobs are still necessary for the monthly ones to work. They gather the accumulation from the daily ones to sum up for the monthly data.

katydidnot replied:
I realize now I misread the acceptance criteria. You are correct, disregard.

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_link_aggregates.py:
```diff
@@ -0,0 +1,184 @@
+                    & start_date_filter
+                )
+                daily_aggregation = (
+                    LinkAggregate.objects.filter(filter_query).exclude(day=0).first()
```
@Amdrel added a `[:1].all()` instead of `first()` in another PR because the latter command added an ORDER BY clause. We should change it to improve performance: https://github.com/WikipediaLibrary/externallinks/pull/416/files#diff-32ce69a9a896fd4220e21debd5d01b071fac7f283ca4eb01d3d52169469454c5R173

mimurawil replied:
Good catch! For this specific case, I changed to use `.exists()` for better performance since I'm not using the record (just checking if it exists)

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_link_aggregates.py:
```diff
@@ -0,0 +1,184 @@
+            year=aggregation_last_day.year,
+            month=aggregation_last_day.month,
+            day=0,
+        ).first()
```
Same as the comment above. Let's see if this change helps with performance.

mimurawil replied:
Thanks! I applied the change. `.all()` seemed redundant here since I'm not chaining any more commands after. From my tests, this still works.

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_pageproject_aggregates.py:
```diff
@@ -0,0 +1,195 @@
+                    & start_date_filter
+                )
+                daily_aggregation = (
+                    PageProjectAggregate.objects.filter(filter_query).exclude(day=0).first()
```
Same performance improvement as in the link aggregate command. Let's see if these changes improve performance

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_pageproject_aggregates.py:
```diff
@@ -0,0 +1,195 @@
+            year=aggregation_last_day.year,
+            month=aggregation_last_day.month,
+            day=0,
+        ).first()
```
Same as above.

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_user_aggregates.py:
```diff
@@ -0,0 +1,192 @@
+                    & start_date_filter
+                )
+                daily_aggregation = (
+                    UserAggregate.objects.filter(filter_query).exclude(day=0).first()
```
Same comment as previous commands.

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_user_aggregates.py:
```diff
@@ -0,0 +1,192 @@
+            year=aggregation_last_day.year,
+            month=aggregation_last_day.month,
+            day=0,
+        ).first()
```
Same comment.

suecarmol reviewed (commented):
Thank you for your work on this! This is my first review pass. I haven't tested this locally, but I have some inline comments.

mimurawil commented:
Sorry, I messed up trying to rebase from master and it auto-closed the PR. Reopening it now.

suecarmol commented:
I've been trying to run the `fill_monthly_pageproject_aggregates` locally, but I am getting an error and the management command dies. Are you getting the same error locally?

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_link_aggregates.py:
```diff
@@ -0,0 +1,192 @@
+
+        logger.info("Monthly LinkAggregate job ended")
+
+    def _get_start_date_filter(self, collection_id=None):
```
Please add documentation to this function.

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_link_aggregates.py:
```diff
@@ -0,0 +1,192 @@
+        """
+        Process all daily aggregations from last month (including) and
+        backwards.
+
```
Please add the `Parameters` and `Returns` documentation to this.

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_link_aggregates.py:
```diff
@@ -0,0 +1,192 @@
+    ):
+        """
+        Saves the monthly aggregation and delete the daily ones.
+
```
Please add the `Parameters` and `Returns` documentation to this.

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_pageproject_aggregates.py:
```diff
@@ -0,0 +1,206 @@
+
+        logger.info("Monthly PageProjectAggregate job ended")
+
+    def _get_start_date_filter(self, collection_id=None):
```
Same as above

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_pageproject_aggregates.py:
```diff
@@ -0,0 +1,206 @@
+    def _process_aggregation(self, collection_id=None, start_date_filter=Q()):
+        """
+        Process all daily aggregations from last month (including) and
+        backwards.
```
Please add the `Parameters` and `Returns` documentation to this.

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_pageproject_aggregates.py:
```diff
@@ -0,0 +1,206 @@
+    ):
+        """
+        Saves the monthly aggregation and delete the daily ones.
+
```
Please add the `Parameters` and `Returns` documentation to this.

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_user_aggregates.py:
```diff
@@ -0,0 +1,200 @@
+
+        logger.info("Monthly UserAggregate job ended")
+
+    def _get_start_date_filter(self, collection_id=None):
```
Please add documentation.

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_user_aggregates.py:
```diff
@@ -0,0 +1,200 @@
+        """
+        Process all daily aggregations from last month (including) and
+        backwards.
+
```
Please add the `Parameters` and `Returns` documentation to this.

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_user_aggregates.py:
```diff
@@ -0,0 +1,200 @@
+    ):
+        """
+        Saves the monthly aggregation and delete the daily ones.
+
```
Please add the `Parameters` and `Returns` documentation to this.

suecarmol commented on extlinks/aggregates/management/commands/fill_monthly_user_aggregates.py:
```diff
@@ -0,0 +1,200 @@
+                        f"Monthly UserAggregate for organisation {prev_item.organisation_id} "
+                        f"collection {prev_item.collection_id} processed successfully"
+                    )
+
```
I am curious to know if you have tried a SUM query similar to what we did on the daily aggregates. Was it too much of a burden to the database?

mimurawil replied:
That's a good point, I haven't considered that. I'm working on testing it as I think this will bring good performance to this script.

mimurawil replied:
Okay, I believe I made the changes to consider SQL SUMs. I did notice a performance improvement, though I also had to apply extra logic in partitioning the huge main query into chunks.

suecarmol reviewed (commented):
Thanks for your work on this! I have a comment I posted above, and some additional changes that are required.

suecarmol commented:
I have been testing the filters in the UI, and it doesn't seem to be working. Also, the calendar widget has disappeared from the inputs. 

**Production**
![Screenshot 2025-02-25 at 12 01 49](https://github.com/user-attachments/assets/6999ec1e-40f4-476d-b61d-ec2be0cf4601)

**Local**
![Screenshot 2025-02-25 at 12 01 54](https://github.com/user-attachments/assets/88891ebd-e0a1-4454-868e-6e7fc22c70ff)

mimurawil commented:
I'm still working on more changes regarding the comments above, for now I pushed:
- Adding documentation to methods
- Reverted the yyyy-mm input type. The issue mentioned above is due to the browser compatibility, I was testing this on Chrome and it does have a calendar pop up for only year & month, but Firefox doesn't support that ☹️. With that in mind, I believe the best solution for now is one of the suggestion in the ticket, leaving the full calendar but always consider full month in the filter
- Adjusted logic to consider the revert above and fixed unit tests according the fixes applied

I didn't get any error messages when running `fill_monthly_pageproject_aggregates` during my tests, though the script takes a long time considering a huge table. I'll make sure to check performance as well.

katydidnot commented:
> I'm still working on more changes regarding the comments above, for now I pushed:
> 
> * Adding documentation to methods
> * Reverted the yyyy-mm input type. The issue mentioned above is due to the browser compatibility, I was testing this on Chrome and it does have a calendar pop up for only year & month, but Firefox doesn't support that ☹️. With that in mind, I believe the best solution for now is one of the suggestion in the ticket, leaving the full calendar but always consider full month in the filter
> * Adjusted logic to consider the revert above and fixed unit tests according the fixes applied
> 
> I didn't get any error messages when running `fill_monthly_pageproject_aggregates` during my tests, though the script takes a long time considering a huge table. I'll make sure to check performance as well.

Hello, thanks for the update! 
I was able to successfully run both the fill_monthly_user_aggregates and fill_monthly_link_aggregates locally. 

However, I am also running into an issue trying to run the fill_monthly_pageproject_aggregates.

mimurawil commented:
Thanks for all the feedback so far!

I made a major logic change hoping to boost performance considerably, though PageProjectAggregate still takes some time to run the monthly data.

In summary, the changes are:
- Monthly aggregation sums totals in its SQL now instead of computing each individual record.
- DB transaction is sectioned into lazy chunks for better memory management
- Added option to run specific year/month
- Added option to run full scan (for historic data)
- Re-running the same month will add to existing totals now
  - It was replacing before. I think it makes more sense to add in case we have new daily aggregations and we want to incorporate to the existing month  
- Default run (no options) will only consider data from last month

Running the `fill_monthly_link_aggregates` goes fast because of the volume in its table (213 aggregations created from ~3k daily aggregations), it took less than 5s:
![Screenshot 2025-02-27 at 5 54 52 PM](https://github.com/user-attachments/assets/5b3ec212-588a-40ad-a194-dbdac2488794)

Running the `fill_monthly_user_aggregates` took more time, around 35m (58k monthly aggregations from ~105k daily ones):
![Screenshot 2025-02-27 at 6 04 26 PM](https://github.com/user-attachments/assets/39417816-7ba1-4e79-8ae3-49cc52412868)


For the `fill_monthly_pageproject_aggregates`, I chose a year/month that had a large amount (over 588k monthly aggregations created, I forgot to count the number of daily aggregations), it took about 4h15m
```
root@c0c290b3ef47:/app# python manage.py fill_monthly_pageproject_aggregates --year-month=2024-01
Monthly PageProjectAggregate job started
Processing data from 2024-01-01 to 2024-01-31
Fetching the main query
Processing batch 1 (size: 10000)
Processing batch 2 (size: 10000)
Processing batch 3 (size: 10000)
Processing batch 4 (size: 10000)
Processing batch 5 (size: 10000)
Processing batch 6 (size: 10000)
...
Processing batch 57 (size: 10000)
Processing batch 58 (size: 10000)
Processing batch 59 (size: 8908)
Processed a total of 588908 monthly aggregations
Monthly PageProjectAggregate job ended
```

mimurawil commented on extlinks/common/helpers.py:
```diff
@@ -152,3 +153,31 @@ def build_queryset_filters(form_data, collection_or_organisations):
     )
+
+
+def batch_iterator(iterable, size=10000):
```
Maybe 10k chunk size is too much. I'm going to test this later with a smaller size to see if we can gain more performance.

mimurawil replied:
Okay, reducing the chunk size indeed improved performance. I ran a few test and found out `500` for PageProjectAggregate is a good number.
- A 10k chunk size was performing around 7m per chunk
- Reducing to 500, I am consistently getting around 14s per chunk (13s ~ 14.5s), which sums to ~4m40s per 10k aggregation

I also tested with `250` and `1000`, but didn't notice much difference in my runs (they were proportionally fast). `500` sounds like a good spot.

As for the other tables (LinkAggregate and UserAggregate), I decided to use `1000` since they are not as big as PageProjectAggregate.

suecarmol replied:
Thanks for all of your work! I'm going to start testing locally now, and I'll let you know if anything is amiss.

mimurawil commented on extlinks/aggregates/management/commands/fill_monthly_pageproject_aggregates.py:
```diff
@@ -0,0 +1,223 @@
+                f"page_name={monthly_aggregation['page_name']}, year={monthly_aggregation['year']}, month={monthly_aggregation['month']}",
+            )
+
+        existing_monthly_aggregate = PageProjectAggregate.objects.filter(
```
This is possibly a performance bottleneck. I thought about storing this in a dictionary so I wound't fetch it for every group, but I got worried about exploding the memory usage

suecarmol replied:
On one of my first Wikilink improvement iterations, I tried doing that, thinking it would improve performance, but it actually got worse. 😢. This was with the linkevent table, so YMMV.

suecarmol commented:
I unfortunately still cannot run the monthly page project aggregate with the `--full-scan` flag. However, I could run the command with the `--year-month` flag, and everything seemed to work as expected.

mimurawil commented:
> I unfortunately still cannot run the monthly page project aggregate with the `--full-scan` flag. However, I could run the command with the `--year-month` flag, and everything seemed to work as expected.

I'm assuming the reason why I can execute it on my machine is because of the capacity, but I see how it could fail to run on the production server since it's a smaller one.
![Screenshot 2025-03-03 at 9 01 32 AM](https://github.com/user-attachments/assets/ecd5b774-ae6e-4aeb-acae-531d380f3b82)

This option may be too expensive, and I can see it taking several days to complete (if the machine can handle the main query). I think the best solution here is to take out the `--full-scan` option and manually run the previous months for the historical data.

katydidnot reviewed (changes requested):
Thanks for your patience and work!

I have successfully ran all monthly cron jobs for all records in the database (I went all the way back to 2019!). I then did a spot check against production. It looks pretty good (no real data anomalies). 

Though there is something broken with the date range. 
After running all of the monthly aggregates for all historical data, if I select a program and then update the date range, I see the top users, but not the top pages or projects. I'll attach a screenshot below: 

![Screenshot 2025-03-06 at 2 52 04 PM](https://github.com/user-attachments/assets/7a3bacf4-beff-4210-89c4-2fed47d0ae68)


As soon as we get that sorted, I'm good to merge this in. Let me know if you have any questions.

katydidnot commented:
@jsnshrmn 
Once this gets merged in, now that we're not doing full table scans, what's the plan for manually running these commands for past aggregates? 

Or are we ok with those hanging around?

jsnshrmn commented:
> @jsnshrmn Once this gets merged in, now that we're not doing full table scans, what's the plan for manually running these commands for past aggregates?
> 
> Or are we ok with those hanging around?

No, we should process the existing data

- the no code way would be for us to just run through all of the year month arguments for each command manually in production
- we could semi automate it with a shell or python script
- we could add back in the "catch it all" option to this pr by having the commands get the oldest date stamp and roll through all of the existing data one month at a time

jsnshrmn reviewed (approved):
LGTM!

jsnshrmn commented:
note that we'll have a followup patch to add a simple script to handle historical data
